File: paper 1/paper1.m
function simulateCpkTable()
    % 設定基本參數
    rng(123);  % 設定隨機種子序確保結果可重現
    Cpk_values = [1, 1.33, 1.5, 2, 2.5, 3];  % Cpk 值的範圍
    n_values = [10, 20, 30, 40, 50];         % 樣本數的範圍
    alpha_values = [0.1, 0.05];              % 顯著水準 alpha
    num_outer_simulations = 10000;           % 外層模擬次數
    num_inner_simulations = 10000;           % 內層模擬次數
    L = 7;                                   % 規格下限
    U = 14;                                  % 規格上限
    mu = 10;                                 % 製程平均值
    d = (U - L) / 2;                         % 規格半寬
    M = (U + L) / 2;                         % 規格中點

    % 預先配置結果陣列空間
    num_Cpk = length(Cpk_values);
    num_n = length(n_values);
    num_alpha = length(alpha_values);
    results = cell(num_Cpk, 1);
    
    % 開始計時
    tic;
    
    % 主要模擬迴圈 (使用平行運算)
    parfor i = 1:num_Cpk
        Cpk = Cpk_values(i);
        sigma = (d - abs(mu - M)) / (3 * Cpk);  % 計算對應的標準差
        temp_results = cell(num_n, num_alpha);
        
        for j = 1:num_n
            n = n_values(j);
            for k = 1:num_alpha
                alpha = alpha_values(k);
                fprintf('Processing Cpk = %.2f, n = %d, alpha = %.2f\n', Cpk, n, alpha);
                sim_result = simulate_Cpk(n, mu, sigma, L, U, Cpk, alpha, ...
                    num_outer_simulations, num_inner_simulations);
                temp_results{j,k} = sim_result;
            end
        end
        results{i} = temp_results;
    end
    
    % 格式化並顯示結果
    formatted_results = format_results(results, Cpk_values, n_values, alpha_values);
    
    % 顯示執行時間
    execution_time = toc;
    fprintf('\nTotal execution time: %.2f seconds\n', execution_time);
end

function result = simulate_Cpk(n, mu, sigma, L, U, Cpk, alpha, num_outer_simulations, num_inner_simulations)
    % 生成隨機樣本
    x = normrnd(mu, sigma, [num_outer_simulations, n]);
    x_bar = mean(x, 2);       % 計算樣本平均值
    s = std(x, 0, 2);         % 計算樣本標準差
    s2 = s.^2;                % 計算樣本變異數
    
    % 計算 Gpk 值
    Gpk_values = zeros(num_outer_simulations, 1);
    for i = 1:num_outer_simulations
        Gpk_values(i) = quantile(calculate_Gpk(x_bar(i), s2(i), n, L, U, num_inner_simulations), alpha);
    end
    
    % 計算其他指標界限
    other_limits = zeros(num_outer_simulations, 4);
    for i = 1:num_outer_simulations
        other_limits(i,:) = calculate_other_limits(x_bar(i), s(i), n, L, U, alpha);
    end
    
    % 合併所有界限
    limits = [Gpk_values, other_limits];
    
    % 計算覆蓋機率和期望值
    coverage = mean(limits <= Cpk, 1);
    expected_values = mean(limits, 1);
    
    % 回傳結構化結果
    result = struct('coverage', round(coverage, 4), ...
                   'expected_values', round(expected_values, 4));
end

function Gpk = calculate_Gpk(x_bar, s2, n, L, U, num_simulations)
    % 計算規格相關參數
    d = (U - L) / 2;
    M = (U + L) / 2;
    
    % 生成隨機數
    Z = randn(num_simulations, 1);              % 標準常態分配
    U2 = chi2rnd(n-1, [num_simulations, 1]);    % 卡方分配
    
    % 計算 T 統計量
    T_mu = x_bar - sqrt((n-1)/n) * (Z * sqrt(s2) / sqrt(n));
    T_sigma2 = s2 * (n-1) ./ U2;
    
    % 計算 Gpk 值
    Gpk = (d - abs(T_mu - M)) ./ (3 * sqrt(T_sigma2));
end

function limits = calculate_other_limits(x_bar, s, n, L, U, alpha)
    % 計算 Cpk 估計值
    Cpk_hat = calculate_Cpk_hat(x_bar, s, L, U);
    z = norminv(1 - alpha);   % 計算標準常態分配的分位數
    
    % 計算各種能力指標界限
    Bpk = Cpk_hat - z * sqrt(1/(9*n) + Cpk_hat^2/(2*(n-1)));
    KHpk = Cpk_hat * (1 - z/sqrt(2*(n-1)));
    Hpk = Cpk_hat - z * sqrt((n-1)/(9*n*(n-3)) + Cpk_hat^2*(1/(2*(n-3))*(1 + 6/(n-1))));
    Npk = sqrt(1-2/(5*(n-1)))*Cpk_hat - z*sqrt(Cpk_hat^2/(2*(n-1)) + 1/(9*n));
    
    limits = [Bpk, KHpk, Hpk, Npk];
end

function Cpk_hat = calculate_Cpk_hat(x_bar, s, L, U)
    % 計算 Cpk 估計值
    Cpk_hat = min((U - x_bar)/(3*s), (x_bar - L)/(3*s));
end

function formatted_table = format_results(results, Cpk_values, n_values, alpha_values)
    % 建立表格標題
    headers = {'Cpk', 'n', '1-α', 'Gpk', 'Bpk', 'KHpk', 'Hpk', 'Npk', ...
              'E(Gpk)', 'E(Bpk)', 'E(KHpk)', 'E(Hpk)', 'E(Npk)'};
    
    % 初始化資料矩陣
    num_rows = length(Cpk_values) * length(n_values) * length(alpha_values);
    data = zeros(num_rows, length(headers));
    row = 1;
    
    % 填充資料矩陣
    for i = 1:length(Cpk_values)
        for j = 1:length(n_values)
            for k = 1:length(alpha_values)
                temp_result = results{i}{j,k};
                data(row,:) = [Cpk_values(i), n_values(j), 1-alpha_values(k), ...
                              temp_result.coverage, temp_result.expected_values];
                row = row + 1;
            end
        end
    end
    
    % 建立並格式化表格
    formatted_table = array2table(data, 'VariableNames', headers);
    
    disp('-------------------------------------------------------------------------------------------------------------------------------');
    disp('                                     Coverage probability                           Expected value');
    disp('-------------------------------------------------------------------------------------------------------------------------------');
    disp(formatted_table);
end

simulateCpkTable()

File: paper 1/paper1.R
library(parallel)     # 用於平行運算
library(knitr)       # 用於產生報表
library(kableExtra)  # 用於美化表格
library(Rcpp)        # 用於整合 C++ 程式碼

# 使用 C++ 實作 Gpk 計算，以提升效能
cppFunction('
NumericVector calculate_Gpk_cpp(double x_bar, double s2, int n, double L, double U, int num_simulations) {
  double d = (U - L) / 2;           // 計算規格半寬
  double M = (U + L) / 2;           // 計算規格中點
  
  NumericVector Z = rnorm(num_simulations);        // 產生標準常態分配的隨機數
  NumericVector U2 = rchisq(num_simulations, n-1); // 產生卡方分配的隨機數
  
  // 計算 T 統計量
  NumericVector T_mu = x_bar - sqrt((n-1.0)/n) * (Z * sqrt(s2) / sqrt(n));
  NumericVector T_sigma2 = s2 * (n-1) / U2;
  
  // 計算並回傳 Gpk 值
  return (d - abs(T_mu - M)) / (3 * sqrt(T_sigma2));
}
')

# 計算 Cpk 值的函數
calculate_Cpk <- function(mu, sigma, L, U) {
  min((U - mu) / (3 * sigma), (mu - L) / (3 * sigma))
}

# 計算其他能力指標界限的函數
calculate_other_limits <- function(x_bar, s, n, L, U, alpha) {
  Cpk_hat <- calculate_Cpk(x_bar, s, L, U)
  z <- qnorm(1 - alpha)  # 計算標準常態分配的分位數
  
  # 將公式 key 入
  Bpk <- Cpk_hat - z * sqrt(1 / (9 * n) + Cpk_hat^2 / (2 * (n - 1)))
  KHpk <- Cpk_hat * (1 - z / sqrt(2 * (n - 1)))
  Hpk <- Cpk_hat - z * sqrt((n - 1) / (9 * n * (n - 3)) + Cpk_hat^2 * (1 / (2 * (n - 3)) * (1 + 6 / (n - 1))))
  Npk <- sqrt(1 - 2 / (5 * (n - 1))) * Cpk_hat - z * sqrt(Cpk_hat^2 / (2 * (n - 1)) + 1 / (9 * n))
  
  c(Bpk, KHpk, Hpk, Npk)
}

# 模擬計算 Cpk 的函數
simulate_Cpk <- function(n, mu, sigma, L, U, Cpk, alpha, num_outer_simulations, num_inner_simulations) {
  true_Cpk <- Cpk
  
  # 生成隨機樣本
  x <- matrix(rnorm(n * num_outer_simulations, mu, sigma), ncol = n)
  x_bar <- rowMeans(x)        # 計算樣本平均值
  s <- apply(x, 1, sd)        # 計算樣本標準差
  s2 <- s^2                   # 計算樣本變異數
  
  # 計算 Gpk 值
  Gpk_values <- sapply(seq_len(num_outer_simulations), function(i) {
    quantile(calculate_Gpk_cpp(x_bar[i], s2[i], n, L, U, num_inner_simulations), alpha)
  })
  
  # 計算其他指標值
  other_limits <- t(sapply(seq_len(num_outer_simulations), function(i) {
    calculate_other_limits(x_bar[i], s[i], n, L, U, alpha)
  }))
  
  limits <- cbind(Gpk_values, other_limits)
  
  # 計算覆蓋機率和期望值
  coverage <- colMeans(limits <= true_Cpk)
  expected_values <- colMeans(limits)
  
  list(coverage = round(coverage, 4), expected_values = round(expected_values, 4))
}

format_results <- function(results) {
  # 將巢狀列表轉換為資料框
  formatted_output <- do.call(rbind, lapply(names(results), function(Cpk) {
    do.call(rbind, lapply(names(results[[Cpk]]), function(n) {
      do.call(rbind, lapply(names(results[[Cpk]][[n]]), function(alpha) {
        c(Cpk, n, 1-as.numeric(alpha),
          results[[Cpk]][[n]][[alpha]]$coverage,
          results[[Cpk]][[n]][[alpha]]$expected_values)
      }))
    }))
  }))
  
  # 設定欄位名稱
  formatted_output <- as.data.frame(formatted_output)
  colnames(formatted_output) <- c("Cpk", "n", "1-α", 
                                  "Gpk", "Bpk", "KHpk", "Hpk", "Npk",
                                  "E(Gpk)", "E(Bpk)", "E(KHpk)", "E(Hpk)", "E(Npk)")
  
  # 轉換數值型態
  formatted_output$Cpk <- as.numeric(as.character(formatted_output$Cpk))
  formatted_output$n <- as.numeric(as.character(formatted_output$n))
  formatted_output$`1-α` <- as.numeric(as.character(formatted_output$`1-α`))
  
  # 四捨五入數值欄位
  numeric_cols <- 4:13
  formatted_output[numeric_cols] <- lapply(formatted_output[numeric_cols], 
                                           function(x) round(as.numeric(as.character(x)), 4))
  
  return(formatted_output)
}

# 設定隨機種子序，確保結果可重現
set.seed(123)  

# 設定模擬參數
Cpk_values <- c(1, 1.33, 1.5, 2, 2.5, 3)
n_values <- c(10, 20, 30, 40, 50)
alpha_values <- c(0.1, 0.05)
num_outer_simulations <- 10000
num_inner_simulations <- 10000
L <- 7    # 規格下限
U <- 14   # 規格上限
mu <- 10  # 製程平均值
d <- (U - L) / 2    # 規格半寬
M <- (U + L) / 2    # 規格中點

start_time <- Sys.time()  # 記錄開始時間

# 執行平行運算模擬
results <- mclapply(Cpk_values, function(Cpk) {
  sigma <- (d - abs(mu - M)) / (3 * Cpk)  # 計算對應的標準差
  
  lapply(n_values, function(n) {
    lapply(alpha_values, function(alpha) {
      cat(sprintf("Processing Cpk = %.2f, n = %d, alpha = %.2f\n", Cpk, n, alpha))
      simulate_Cpk(n, mu, sigma, L, U, Cpk, alpha, num_outer_simulations, num_inner_simulations)
    })
  })
}, mc.cores = detectCores())  # 使用所有可用的 CPU 核心

# 設定結果列表的名稱
names(results) <- as.character(Cpk_values)
for (i in seq_along(results)) {
  names(results[[i]]) <- as.character(n_values)
  for (j in seq_along(results[[i]])) {
    names(results[[i]][[j]]) <- as.character(alpha_values)
  }
}

end_time <- Sys.time()  # 記錄結束時間
total_time <- difftime(end_time, start_time, units = "secs")

# 格式化並顯示結果
formatted_results <- format_results(results)

# 修改欄位名稱以增加間距
names(formatted_results)[3] <- "1-α   "
names(formatted_results)[8] <- "Npk   "

# 使用 kable 產生美化的表格
print(kable(formatted_results, 
            format = "pipe",
            digits = 4,
            align = c('r', 'r', 'r', rep('r', 5), rep('r', 5))) %>%
        add_header_above(c(" " = 3, 
                           "Coverage probability" = 5,
                           "Expected value" = 5)) %>%
        add_header_above(c(" " = 13)) %>%
        kable_styling(full_width = F))

# 印出執行時間
cat(sprintf("\nTotal execution time: %.2f 秒\n", total_time))